### ISSUE TITLE
Segmentation Fault in newest version

### ISSUE THREAD

**Issue by agtoever** [stevens68/twixtbot-ui#85]
Using the latest version (8bb8915), I get consistently a segmentation fault after the following steps:

- Start twixtbot-ui
- Set all settings to default
- Turn on "smart root" for both players
- In the main window, enable auto play for both players
- Before the end of the game, twixtbot-ui crashes.

```
Fatal Python error: Segmentation fault

Current thread 0x000070000411b000 (most recent call first):
  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/_backend_tk.py", line 77 in blit
  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/backend_tkagg.py", line 10 in draw
  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/src/plot.py", line 35 in update
  File "./tbui.py", line 229 in update_evals
  File "./tbui.py", line 292 in update_after_move
  File "./tbui.py", line 380 in bot_move
  File "/usr/local/Cellar/python@3.8/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/threading.py", line 870 in run
  File "/usr/local/Cellar/python@3.8/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/local/Cellar/python@3.8/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x0000000116b29e00 (most recent call first):
  File "/usr/local/Cellar/python@3.8/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 1320 in update_idletasks
  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/backend_tkagg.py", line 11 in draw
  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/_backend_tk.py", line 253 in idle_draw
  File "/usr/local/Cellar/python@3.8/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 814 in callit
  File "/usr/local/Cellar/python@3.8/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 1892 in __call__
  File "/usr/local/Cellar/python@3.8/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 1320 in update_idletasks
  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/backend_tkagg.py", line 11 in draw
  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/_backend_tk.py", line 242 in resize
  File "/usr/local/Cellar/python@3.8/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 1892 in __call__
  File "/usr/local/Cellar/python@3.8/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 1429 in mainloop
  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/PySimpleGUI/PySimpleGUI.py", line 8472 in _read
  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/PySimpleGUI/PySimpleGUI.py", line 8354 in read
  File "./tbui.py", line 590 in get_event
  File "./tbui.py", line 785 in main
  File "./tbui.py", line 801 in <module>
[1]    39010 segmentation fault  ./tbui.py

```

**Comment by stevens68** [comment]
….thanks for letting me know. Just noticed your email. Strange. It happens in plot.py that I didn’t touch. I will have a look.(I’ve merged another PR meanwhile) Regards, stevens68  PS: ..and thanks for introducing the logger!  From: Albert ten OeverSent: Saturday, 15 May 2021 22:30To: stevens68/twixtbot-uiCc: SubscribedSubject: [stevens68/twixtbot-ui] Segmentation Fault in newest version (#85) Using the latest version, I get consistently a segmentation fault after the following steps:Start twixtbot-uiSet all settings to defaultTurn on "smart root" for both playersIn the main window, enable auto play for both playersBefore the end of the game, twixtbot-ui crashes.Fatal Python error: Segmentation fault Current thread 0x000070000411b000 (most recent call first):  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/_backend_tk.py", line 77 in blit  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/backend_tkagg.py", line 10 in draw  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/src/plot.py", line 35 in update  File "./tbui.py", line 229 in update_evals  File "./tbui.py", line 292 in update_after_move  File "./tbui.py", line 380 in bot_move  File ***@***.***/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/threading.py", line 870 in run  File ***@***.***/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/threading.py", line 932 in _bootstrap_inner  File ***@***.***/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/threading.py", line 890 in _bootstrap Thread 0x0000000116b29e00 (most recent call first):  File ***@***.***/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 1320 in update_idletasks  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/backend_tkagg.py", line 11 in draw  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/_backend_tk.py", line 253 in idle_draw  File ***@***.***/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 814 in callit  File ***@***.***/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 1892 in __call__  File ***@***.***/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 1320 in update_idletasks  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/backend_tkagg.py", line 11 in draw  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/matplotlib/backends/_backend_tk.py", line 242 in resize  File ***@***.***/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 1892 in __call__  File ***@***.***/3.8.9/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py", line 1429 in mainloop  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/PySimpleGUI/PySimpleGUI.py", line 8472 in _read  File "/Users/agtoever/Dropbox/Eclipse workspace/twixtbot-ui/venv/lib/python3.8/site-packages/PySimpleGUI/PySimpleGUI.py", line 8354 in read  File "./tbui.py", line 590 in get_event  File "./tbui.py", line 785 in main  File "./tbui.py", line 801 in <module>[1]    39010 segmentation fault  ./tbui.py —You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by stevens68** [comment]
Couldn't reproduce it on my system (3.8.9 and 3.7.9) 

It seems it's because I moved execution of the move AND some UI updates into the thread.
I missed that tkinter doesn't like to be called from different threads (it's not supported)

Will move the UI updates out of  bot_move(). 
You might want to uncomment lines 376, 380 and check if it still crashes.

            #self.get_control(ct.K_SPINNER).Update(visible=False)
            if not self.bot_event.is_set() or self.bot_event.get_context() == ct.ACCEPT_EVENT:
                # bot has not been cancelled (but is finished or accepted)
                self.execute_move(response["moves"][0])
                #self.update_after_move(False)

**Comment by agtoever** [comment]
Yes. Disabling self.update_after_move() prevents it from crashing on my machine.

I think the underlying problem is that the TwixtbotUI class has become a [God-class](https://wiki.c2.com/?GodClass) and that game logic, GUI handling and GUI functionality handling are spread over different files and classes. We can (and should) fix this segmentation fault bug, caused by doing UI updates from a thread that should only do bot calculations, but in another issue, I think we should seriously consider refactoring the TwixtbotUI class (and along with that some other classes).

To further demonstrate this point, I used [pyan3](https://pypi.org/project/pyan3/) to create the following call graph. This is *only* tbui.py(!)

![tbui_graphviz](https://user-images.githubusercontent.com/13528099/118868474-85191580-b8e4-11eb-8f25-424e3b9375b4.png)

**Comment by stevens68** [comment]
Would be great if you could test the latest version; Hope the seg fault is fixed with PR#87. Let me know if not.

Yes, TwixtbotUI class has indeed become quite big and could use some refactoring.

**Comment by agtoever** [comment]
Solved in PR#87
